File: generators/zml/state.zig
//! D4's state laws, implemented once (Z08).
//!
//! Every ZML model today hand-writes its own KV cache. D4 declares the law, the access
//! geometry and the payload of every state, so the three laws are three implementations
//! for the whole corpus — a primitive never allocates or shapes a state, it reads and
//! appends.
//!
//! In a traced graph a state is functional: buffers arrive as operands and leave as
//! results.
//!
//! **The model exposes a state; the serving application decides how to hold it.**
//!
//! The line between the two is the whole point, so it is worth stating exactly:
//!
//! | From the document (D4) | From the serving application |
//! |---|---|
//! | the law, the access geometry, the span | how many positions (capacity) |
//! | the payload: components, axes, extents, dtype | one buffer per identity, one per family, or pages |
//! | that there is one identity per layer | which portion of a shared buffer each occurrence consumes |
//!
//! **Geometry is inferred from the model and nowhere else** — `instanceOf` reads it out
//! of D4's `payload`, `law` and `span`, and `packableWith` refuses to put two states in
//! one buffer unless the document says their payloads agree. A serving choice may
//! rearrange states; it may never decide their shape.
//!
//! The layout, by contrast, is the serving application's, exactly like a block size or a
//! kernel selection — so the portion an occurrence consumes reaches its primitive
//! through the **opaque physical channel** (`generators/CAPABILITIES.md`), never through
//! the contract's arguments. A primitive asks this handle for its own view and never
//! learns how many neighbours share the buffer, nor in what order, nor whether the view
//! it gets is a slice of one allocation or gathered from pages. That is what lets the
//! same primitive serve a naive contiguous cache and a paged one.

const std = @import("std");

const zml = @import("zml");

const graph = @import("graph.zig");

pub const Error = error{
    /// A law D4 declares and this generator does not implement. A finding, not a
    /// workaround: the enum is closed, so a fourth law is a language matter.
    UnimplementedLaw,
    UnimplementedAccess,
    /// `window` needs the span D4 gives it.
    MissingSpan,
};

pub const Law = enum { append, window, fixed };
pub const Access = enum { logical_position, ring, aggregate, selected };

pub fn lawOf(name: []const u8) Error!Law {
    return std.meta.stringToEnum(Law, name) orelse Error.UnimplementedLaw;
}

pub fn accessOf(name: []const u8) Error!Access {
    return std.meta.stringToEnum(Access, name) orelse Error.UnimplementedAccess;
}

/// One payload component's buffer. `[positions, *payload]` for a growing or bounded law
/// (§4.3: a payload is per position for append/window, the whole state for fixed), with
/// a leading `[members]` axis when the serving layout packs a family together.
pub const Component = struct {
    name: []const u8,
    shape: zml.Shape,
};

/// The buffers one *layout* needs: the D4 identities it packs, and their components.
/// One identity per member when nothing is packed.
pub const Instance = struct {
    /// The D4 identities this layout holds, in member order. A dump names its files by
    /// these, never by the layout: what the runtime chose to pack is its own business,
    /// and what leaves the generator is what the document declared.
    identities: []const []const u8,
    law: Law,
    access: Access,
    /// How many D4 identities share these buffers. 1 when each has its own.
    members: usize,
    components: []const Component,

    /// Whether two D4 states can share one buffer: same law, same access, same payload.
    /// Identity is not consulted — that they are one family is D4's business, and this
    /// is only asking whether one allocation can hold both.
    pub fn packableWith(self: Instance, s: graph.State, compute: zml.DataType) bool {
        if (self.law != (lawOf(s.law) catch return false)) return false;
        if (self.access != (accessOf(s.access) catch return false)) return false;
        if (self.components.len != s.payload.len) return false;
        for (self.components, s.payload) |c, p| {
            if (!std.mem.eql(u8, c.name, p.component)) return false;
            if (p.shape.len + 2 != c.shape.rank()) return false;
            for (p.shape, 0..) |axis, i| {
                if (c.shape.dim(@as(i64, @intCast(i)) + 2) != axis.extent) return false;
            }
            if (c.shape.dtype() != compute) return false;
        }
        return true;
    }
};

/// The buffers D4 asks for, under a capacity the deployment fixed and a layout the
/// serving application chose. Capacity is deployment intent, not a document fact: D4
/// gives the bytes per position, §7 says how many positions is the runtime's business.
pub fn instanceOf(
    allocator: std.mem.Allocator,
    s: graph.State,
    capacity: i64,
    compute: zml.DataType,
) !Instance {
    const law = try lawOf(s.law);
    const access = try accessOf(s.access);

    const positions: i64 = switch (law) {
        .append => capacity,
        .window => s.span orelse return Error.MissingSpan,
        .fixed => 1,
    };

    const components = try allocator.alloc(Component, s.payload.len);
    for (s.payload, components) |p, *c| {
        // [members, positions, *payload]; the members axis grows as identities join.
        var sh = zml.Shape.init(.{}, compute);
        sh = sh.appendDim(1, null);
        sh = sh.appendDim(positions, null);
        for (p.shape) |axis| sh = sh.appendDim(axis.extent, null);
        c.* = .{ .name = p.component, .shape = sh };
    }
    const identities = try allocator.alloc([]const u8, 1);
    identities[0] = s.identity;
    return .{ .identities = identities, .law = law, .access = access, .members = 1, .components = components };
}

/// One more identity joins this layout's buffers.
pub fn addMember(allocator: std.mem.Allocator, instance: *Instance, identity: []const u8) !void {
    const grown = try allocator.alloc(Component, instance.components.len);
    for (instance.components, grown) |c, *g| {
        g.* = .{ .name = c.name, .shape = c.shape.setDim(0, c.shape.dim(0) + 1) };
    }
    instance.components = grown;

    const identities = try allocator.alloc([]const u8, instance.identities.len + 1);
    @memcpy(identities[0..instance.identities.len], instance.identities);
    identities[instance.identities.len] = identity;
    instance.identities = identities;
    instance.members += 1;
}

/// What a primitive is handed for one state: its buffers, where this invocation's
/// elements sit in them, and — from the opaque channel — which portion it owns.
pub const Handle = struct {
    law: Law,
    access: Access,
    /// The whole buffers, members included. A primitive never indexes these directly.
    buffers: []const zml.Tensor,
    names: []const []const u8,
    /// The portion of the packed buffers this occurrence consumes. A **physical
    /// parameter**: which layer sits where is the serving layout's business, opaque to
    /// the language and passed beside the contract's arguments, never merged into them.
    member: i64,
    /// The logical position of this invocation's first element.
    start: zml.Tensor,
    /// How many elements this invocation carries.
    elements: i64,

    /// This occurrence's own view of a component — `[positions, *payload]`, with the
    /// packing gone.
    pub fn get(self: Handle, name: []const u8) ?zml.Tensor {
        for (self.names, self.buffers) |n, b| {
            // `.single` drops the axis: the primitive sees `[positions, *payload]` and
            // nothing of the layout that produced it.
            if (std.mem.eql(u8, n, name)) return b.slice(0, .single(self.member));
        }
        return null;
    }

    /// The buffers with this invocation's values written into this occurrence's
    /// portion. For `append`, elements land at `start`, contiguously — which is what
    /// the reference's cursor does.
    pub fn append(self: Handle, allocator: std.mem.Allocator, values: []const zml.Tensor) ![]zml.Tensor {
        std.debug.assert(values.len == self.buffers.len);
        if (self.law != .append) return Error.UnimplementedLaw;

        const out = try allocator.alloc(zml.Tensor, self.buffers.len);
        for (self.buffers, values, out) |buffer, value, *updated| {
            const offsets = try allocator.alloc(zml.Tensor, buffer.rank());
            defer allocator.free(offsets);
            offsets[0] = zml.Tensor.scalar(self.member, .i32);
            offsets[1] = self.start.convert(.i32);
            for (offsets[2..]) |*o| o.* = zml.Tensor.scalar(0, .i32);

            // The update carries the members axis as a single position of its own.
            const update = value.convert(buffer.dtype()).insertAxes(0, .{.portion});
            updated.* = buffer.dynamicUpdateSlice(offsets, update);
        }
        return out;
    }

    /// The same handle over buffers that have just been written — so a primitive reads
    /// its own portion of what it appended without knowing the layout.
    pub fn after(self: Handle, buffers: []const zml.Tensor) Handle {
        var next = self;
        next.buffers = buffers;
        return next;
    }

    /// How many positions of this occurrence's portion hold written values.
    pub fn length(self: Handle) zml.Tensor {
        return self.start.addConstant(self.elements);
    }
};

File: generators/reference/state.py
"""State instances: one per D4 identity for one (session, branch), each law
implemented once under a declared capacity (R04, R07).

    append   buffer [capacity, *shape], a cursor; read() -> (buffers, length)
    window   buffer [span, *shape] as a ring; read() -> chronological, zero-padded
    fixed    one tensor per component; read()/write()

Kernels read a whole buffer plus a length and mask beyond it — the same code
eager or compiled.
"""
import torch


LAWS = ('append', 'window', 'fixed')                 # the state laws implemented below
ACCESS = ('logical_position', 'ring', 'aggregate')    # the access geometries the laws realise


class Refusal(Exception):
    pass


class StateInstance:
    def __init__(self, entry, capacity, device, dtype):
        self.identity = entry['identity']
        self.law = entry['law']
        self.access = entry['access']
        self.span = entry.get('span')
        self.capacity = capacity if self.law == 'append' else (self.span if self.law == 'window' else None)
        if self.law not in LAWS:
            raise Refusal(f"{self.identity}: unknown state law '{self.law}'")
        self.components = {}
        for p in entry['payload']:
            # A payload is one position for `append` and `window`, the state for `fixed` (§4.3).
            # Buffers are element-first, [positions, *payload]: a window's ring is [span, *payload].
            shape = [a['extent'] for a in p['shape']]
            full = ([self.capacity] + shape) if self.capacity is not None else shape
            self.components[p['component']] = torch.zeros(full, device=device, dtype=dtype)
        self.length = 0
        self.cursor = 0                                            # window: the next write position

    def read(self):
        if self.law == 'fixed':
            return self.components, None
        if self.law == 'append':
            return self.components, self.length
        n = min(self.length, self.span)
        out = {}
        for c, buf in self.components.items():
            if self.length < self.span:
                chrono = buf[:n]                                   # written from 0, cursor == n
            else:
                chrono = torch.cat([buf[self.cursor:], buf[:self.cursor]], dim=0)
            if n < self.span:
                pad = torch.zeros((self.span - n,) + tuple(buf.shape[1:]), device=buf.device, dtype=buf.dtype)
                chrono = torch.cat([pad, chrono], dim=0)
            out[c] = chrono
        return out, n

    def append(self, values):
        m = next(iter(values.values())).shape[0]
        if self.law == 'append':
            if self.length + m > self.capacity:
                raise Refusal(f"{self.identity}: {self.length + m} positions exceed the capacity {self.capacity}")
            for c, x in values.items():
                self.components[c][self.length:self.length + m] = x.to(self.components[c].dtype)
        elif self.law == 'window':
            idx = (self.cursor + torch.arange(m, device=next(iter(values.values())).device)) % self.span
            for c, x in values.items():
                if m >= self.span:
                    self.components[c].copy_(x[-self.span:].to(self.components[c].dtype))
                else:
                    self.components[c][idx] = x.to(self.components[c].dtype)
            self.cursor = 0 if m >= self.span else (self.cursor + m) % self.span
        else:
            raise Refusal(f"{self.identity}: append on a fixed state")
        self.length += m

    def write(self, values):
        if self.law != 'fixed':
            raise Refusal(f"{self.identity}: write on a {self.law} state")
        for c, x in values.items():
            self.components[c] = x.to(self.components[c].dtype)
        self.length = 1

    def reset(self):
        for c in self.components:
            self.components[c].zero_()
        self.length = 0
        self.cursor = 0

    def clone(self):
        """An independent copy: the buffers, the length and the cursor as they stand."""
        other = StateInstance.__new__(StateInstance)
        other.__dict__.update(self.__dict__)
        other.components = {c: buf.clone() for c, buf in self.components.items()}
        return other

    def truncate(self, at):
        """An `append` state cut back to its first `at` positions (sharing `by_position`, §4.3):
        what a session forked at `at` may read of its parent. The tail is zeroed, never read."""
        if self.law != 'append':
            raise Refusal(f"{self.identity}: only an append state holds positions to truncate to")
        if at > self.length:
            raise Refusal(f"{self.identity}: position {at} is beyond the {self.length} held")
        for c, buf in self.components.items():
            buf[at:self.length].zero_()
        self.length = at


def allocate(graph, capacity, device, dtype):
    return {ident: StateInstance(entry, capacity, device, dtype) for ident, entry in graph.states.items()}


def bytes_per_position(graph):
    return graph.d4_totals.get('append_bytes_per_cached_position', 0)
